### ISSUE TITLE
Problems with deliverables when using the full_fp16 option in lora learning

### ISSUE THREAD

**Issue by ytoaa** [kohya-ss/sd-scripts#1914]
If you use the full_fp16 option, the learning will proceed, but there is no change at all when you output it as a result afterwards. I tried turning Laura on/off on the same model, same seed, but the same result was output. So if I turn off the full_fp16 option and turn on the "mixed_precision = fp16" option to proceed, the result will be normal. It seems that there is simply a problem with the full_fp16 option. I tried it on the default and sd3 branches, but the problem happened the same.

Below are the options used for learning.
The problem is the same even if you remove the locon and lora+ options from the settings below and try.

This might be a problem that only happened in my environment, so I tried with the same option in other people's environments, but everyone had the same problem.

{
  "LoRA_type": "LyCORIS/LoCon",
  "LyCORIS_preset": "full",
  "adaptive_noise_scale": 0,
  "additional_parameters": "",
  "ae": "",
  "apply_t5_attn_mask": false,
  "async_upload": false,
  "block_alphas": "",
  "block_dims": "",
  "block_lr_zero_threshold": "",
  "blocks_to_swap": 0,
  "bucket_no_upscale": true,
  "bucket_reso_steps": 64,
  "bypass_mode": false,
  "cache_latents": true,
  "cache_latents_to_disk": true,
  "caption_dropout_every_n_epochs": 0,
  "caption_dropout_rate": 0,
  "caption_extension": ".txt",
  "clip_g": "",
  "clip_g_dropout_rate": 0,
  "clip_l": "",
  "clip_skip": 2,
  "color_aug": false,
  "constrain": 0,
  "conv_alpha": 32,
  "conv_block_alphas": "",
  "conv_block_dims": "",
  "conv_dim": 32,
  "cpu_offload_checkpointing": false,
  "dataset_config": "",
  "debiased_estimation_loss": true,
  "decompose_both": false,
  "dim_from_weights": false,
  "discrete_flow_shift": 3,
  "dora_wd": false,
  "double_blocks_to_swap": 0,
  "down_lr_weight": "",
  "dynamo_backend": "no",
  "dynamo_mode": "default",
  "dynamo_use_dynamic": false,
  "dynamo_use_fullgraph": false,
  "enable_all_linear": false,
  "enable_bucket": true,
  "epoch": 10,
  "extra_accelerate_launch_args": "",
  "factor": -1,
  "flip_aug": false,
  "flux1_cache_text_encoder_outputs": false,
  "flux1_cache_text_encoder_outputs_to_disk": false,
  "flux1_checkbox": false,
  "fp8_base": false,
  "fp8_base_unet": false,
  "full_bf16": false,
  "full_fp16": true,
  "gpu_ids": "",
  "gradient_accumulation_steps": 1,
  "gradient_checkpointing": true,
  "guidance_scale": 3.5,
  "highvram": false,
  "huber_c": 0.1,
  "huber_scale": 1,
  "huber_schedule": "snr",
  "huggingface_path_in_repo": "",
  "huggingface_repo_id": "",
  "huggingface_repo_type": "",
  "huggingface_repo_visibility": "",
  "huggingface_token": "",
  "img_attn_dim": "",
  "img_mlp_dim": "",
  "img_mod_dim": "",
  "in_dims": "",
  "ip_noise_gamma": 0.05,
  "ip_noise_gamma_random_strength": false,
  "keep_tokens": 0,
  "learning_rate": 1,
  "log_config": false,
  "log_tracker_config": "",
  "log_tracker_name": "",
  "log_with": "",
  "logging_dir": "Q:/chro/output/log",
  "logit_mean": 0,
  "logit_std": 1,
  "loraplus_lr_ratio": 16,
  "loraplus_text_encoder_lr_ratio": 16,
  "loraplus_unet_lr_ratio": 16,
  "loss_type": "huber",
  "lowvram": false,
  "lr_scheduler": "cosine",
  "lr_scheduler_args": "",
  "lr_scheduler_num_cycles": 2,
  "lr_scheduler_power": 1,
  "lr_scheduler_type": "",
  "lr_warmup": 5,
  "lr_warmup_steps": 0,
  "main_process_port": 0,
  "masked_loss": false,
  "max_bucket_reso": 2048,
  "max_data_loader_n_workers": 2,
  "max_grad_norm": 1,
  "max_resolution": "1024,1024",
  "max_timestep": 1000,
  "max_token_length": 225,
  "max_train_epochs": 0,
  "max_train_steps": 0,
  "mem_eff_attn": false,
  "mem_eff_save": false,
  "metadata_author": "",
  "metadata_description": "",
  "metadata_license": "",
  "metadata_tags": "",
  "metadata_title": "",
  "mid_lr_weight": "",
  "min_bucket_reso": 64,
  "min_snr_gamma": 5,
  "min_timestep": 0,
  "mixed_precision": "fp16",
  "mode_scale": 1.29,
  "model_list": "custom",
  "model_prediction_type": "sigma_scaled",
  "module_dropout": 0,
  "multi_gpu": false,
  "multires_noise_discount": 0,
  "multires_noise_iterations": 0,
  "network_alpha": 32,
  "network_dim": 32,
  "network_dropout": 0,
  "network_weights": "",
  "noise_offset": 0.0357,
  "noise_offset_random_strength": false,
  "noise_offset_type": "Original",
  "num_cpu_threads_per_process": 2,
  "num_machines": 1,
  "num_processes": 1,
  "optimizer": "Prodigy",
  "optimizer_args": "d_coef=1 safeguard_warmup=True decouple=True weight_decay=0.01 betas=0.9,0.99 use_bias_correction=True",
  "output_dir": "Q:/chro/output",
  "output_name": "kaede_to_suzu_local_v1",
  "persistent_data_loader_workers": true,
  "pos_emb_random_crop_rate": 0,
  "pretrained_model_name_or_path": "Q:/ComfyUI/models/checkpoints/Ilxl/illustriousXL_v01.safetensors",
  "prior_loss_weight": 1,
  "random_crop": false,
  "rank_dropout": 0,
  "rank_dropout_scale": false,
  "reg_data_dir": "",
  "rescaled": false,
  "resume": "",
  "resume_from_huggingface": "",
  "sample_every_n_epochs": 0,
  "sample_every_n_steps": 0,
  "sample_prompts": "",
  "sample_sampler": "euler_a",
  "save_as_bool": false,
  "save_clip": false,
  "save_every_n_epochs": 1,
  "save_every_n_steps": 0,
  "save_last_n_epochs": 0,
  "save_last_n_epochs_state": 0,
  "save_last_n_steps": 0,
  "save_last_n_steps_state": 0,
  "save_model_as": "safetensors",
  "save_precision": "fp16",
  "save_state": false,
  "save_state_on_train_end": false,
  "save_state_to_huggingface": false,
  "save_t5xxl": false,
  "scale_v_pred_loss_like_noise_pred": false,
  "scale_weight_norms": 0,
  "sd3_cache_text_encoder_outputs": false,
  "sd3_cache_text_encoder_outputs_to_disk": false,
  "sd3_checkbox": false,
  "sd3_clip_l": "",
  "sd3_clip_l_dropout_rate": 0,
  "sd3_disable_mmap_load_safetensors": false,
  "sd3_enable_scaled_pos_embed": false,
  "sd3_fused_backward_pass": false,
  "sd3_t5_dropout_rate": 0,
  "sd3_t5xxl": "",
  "sd3_text_encoder_batch_size": 1,
  "sdxl": true,
  "sdxl_cache_text_encoder_outputs": false,
  "sdxl_no_half_vae": true,
  "seed": 0,
  "shuffle_caption": true,
  "single_blocks_to_swap": 0,
  "single_dim": "",
  "single_mod_dim": "",
  "skip_cache_check": false,
  "split_mode": false,
  "split_qkv": false,
  "stop_text_encoder_training": 0,
  "t5xxl": "",
  "t5xxl_device": "",
  "t5xxl_dtype": "bf16",
  "t5xxl_lr": 0,
  "t5xxl_max_token_length": 512,
  "text_encoder_lr": 1,
  "timestep_sampling": "sigma",
  "train_batch_size": 4,
  "train_blocks": "all",
  "train_data_dir": "Q:/chro/kaede_suzu_style/img",
  "train_double_block_indices": "all",
  "train_norm": true,
  "train_on_input": true,
  "train_single_block_indices": "all",
  "train_t5xxl": false,
  "training_comment": "",
  "txt_attn_dim": "",
  "txt_mlp_dim": "",
  "txt_mod_dim": "",
  "unet_lr": 1,
  "unit": 1,
  "up_lr_weight": "",
  "use_cp": false,
  "use_scalar": true,
  "use_tucker": false,
  "v2": false,
  "v_parameterization": false,
  "v_pred_like_loss": 0,
  "vae": "",
  "vae_batch_size": 0,
  "wandb_api_key": "",
  "wandb_run_name": "",
  "weighted_captions": false,
  "weighting_scheme": "logit_normal",
  "xformers": "xformers"
}

**Comment by DKnight54** [comment]
@kohya-ss I would note that this issue only occurs if using adaptive optimizers such as prodigy and adafactor when they are set to dynamic learning rates.

**Comment by ytoaa** [comment]
I tried adamw, adamw8bit.
Adamw was not able to learn. (nan)
The adamw8bit was able to learn, but there were only minor changes.
I don't know if there is a full_fp16 and an optimizer that works fine.

**Comment by DKnight54** [comment]
@ytoaa the following settings work  with adafactor optimizer (Fixed LR)
```
optimizer_type = "adafactor"
optimizer_args = [ "scale_parameter=False", "relative_step=False", "warmup_init=False" ]
lr_scheduler = "constant_with_warmup"
lr_warmup_steps = 100
learning_rate = 4e-7 # SDXL original learning rate
```

No promises on the quality of the results though

**Comment by ytoaa** [comment]
> [@ytoaa](https://github.com/ytoaa) the following settings work with adafactor optimizer (Fixed LR)
> 
> ```
> optimizer_type = "adafactor"
> optimizer_args = [ "scale_parameter=False", "relative_step=False", "warmup_init=False" ]
> lr_scheduler = "constant_with_warmup"
> lr_warmup_steps = 100
> learning_rate = 4e-7 # SDXL original learning rate
> ```
> 
> No promises on the quality of the results though


Thank you, and I'll give it a try.

**Comment by ytoaa** [comment]
If you use "adafactor" and raise the learning rate to around 5e-4, it shows a good result. In fact, I think there are two kinds of optimizers that are possible: "adafactor, Adamw8bit."

**Comment by kekm8-cpu** [comment]
> In fact, there seems to be no reason for the presence of full_fp16 in lora training...

I could be wrong here, but I imagine the benefit would be both training and image generation happening in the same format.
Image generation is in fp16. So after training a lora in bf16, it would need to be cast to fp16 to actually use it.

In casting a lora from bf16 to fp16, you get the worst of both formats. The higher ranges offered by bf16 would be downscaled while the precision of fp16 would not be taken advantage of since the lora was trained in bf16. It's a lose-lose.

I think the best solution would be to cast the entire flux model to bf16 so that loras can run without being recast to fp16.
The problem is that no software I'm aware of is capable of running Flux in bf16.

My guess is that this is because the main (possibly only) difference between the free Flux Dev and Flux Pro is that Flux Pro is formatted in the superior bf16. If people could run, and not just train, Flux in bf16 it would ruin BFL's business model.

I am curious what all of your thoughts are and if it's possible to run recast Flux Dev to bf16 and run it that way locally.
